mathesar_ui/src/component-library/pagination/Pagination.svelte

```diff
@@ -29,6 +29,12 @@
     undefined;
 
   // Total number of pages.
+  //
+  // TODO: @seancolsen says:
+  // > Refactor `pageCount` to no longer be an exported prop. We're exporting it
+  // > just so the parent component can access the calculation done within this
+  // > component. That's an unconventional flow of data. I'd rather do the
```

Comment:
It's unconventional for other frameworks but the whole point of the `bind` keyword is to do exactly this. There are even dedicated readonly `bind` properties that Svelte provides. I'd say this is one of the intended ways of doing things in Svelte.

Reply:
> the whole point of the bind keyword is to do exactly this

I disagree. I would say the point of `bind` is to allow data to flow from child to parent. `bind:value` on an `input` element allows data that _originates_ in the `input` element (from user input) to flow up to the parent. That's great.

This `Pagination` component is different. The data for `pageCount` originates in the parent component as `total` and `pageSize`. The `pageCount` value is simply a derivation of those two values, which `Pagination` never modifies. Thus the source-of-truth data here is being passed _down_. Then a _derivation_ of that data is being passed back _up_. That's weird an unconventional. We have lots of `bind` usage in our codebase which is fine because the source of truth for the data actually originates in the child. But this is the only one I've seen which, I would say, abuses the `bind` pattern. Grappling with this weirdness costed me time during this refactor, which is why I suggested we clean it up.

Reply:
I see your point regarding `bind` keyword.

For the `pageCount` property however, I strongly feel that it should be the child component's property and not be passed down from the parent.

Even though `pageSize` and `total` are passed down, `pageCount` is still a property which is calculated within the child component inorder for it to work as expected. The source of truth for `pageCount` is where the calculation happens and not where it's dependencies are passed from.

We can remove the `bind` keyword here and dispatch an event when `pageCount` changes and use that to update any property on the parent.

Reply:
> We can remove the bind keyword here and dispatch an event when pageCount changes and use that to update any property on the parent.

That would be problematic in the same way.

If I were designing this from scratch, I'd make a `PaginationStats` object like this:

```ts
class PaginationStats {
  readonly recordCount: number;
  readonly pageSize: number;
  readonly pageCount: number;

  constructor({
    recordCount,
    pageSize
  }: {
    recordCount: number,
    pageSize: number,
  }) {
    this.recordCount = recordCount;
    this.pageSize = pageSize;
    this.pageCount = Math.ceil(this.recordCount / this.pageSize);
  }
}
```

Then I'd make a `stats` prop on `Pagination` which accepts an instance of `PaginationStats`.

In our code currently, the `StatusPane` needs access to the `pageCount`. The _only_ way it can get the `pageCount` is to render a `Pagination` component. That's deeply wierd to me. What if we want to take `Pagination` out of `StatusPane` but keep `pageCount` in `StatusPane`? Since `StatusPane` needs access to `pageCount`, `pageCount` should not be coupled to `Pagination`. The design above decouples `pageCount` from `Pagination`.

Reply:
> That would be problematic in the same way.

Not necessarily. `pageCount` would be decoupled since the parent's pageCount and the child's pageCount are different variables. 

> What if we want to take Pagination out of StatusPane but keep pageCount in StatusPane

We'd pass down `pageCount` to` StatusPane`.

> The only way it can get the pageCount is to render a Pagination component.

It's the only `current` way to do it. If we absolutely need to calculate pageCount on the parent, we'd best expose a util method in the module part of the Pagination component. That way any file can do:
`import { calculatePageCount } from 'Pagination.svelte'`

> If I were designing this from scratch, I'd make a PaginationStats object like this
> Then I'd make a stats prop on Pagination which accepts an instance of PaginationStats.

This is a design pattern I'd go with only when it is absolutely essential.

It makes the Pagination component reliant on the PaginationStats object, which means that anyone wanting to use the Pagination component needs to create an instance of the PaginationStats class. Reactive updates of `page` and `pageSize` would always need creating a new object, leading to boilerplate code wherever we use this. It's one of the annoying patterns/aspects of React that made me like Svelte.

We can solve this simply by exporting some util methods, as I mentioned above.

Reply:
The utils pattern you suggest above is great too. It's certainly less code than my example. The calculation function would get called more times than in my proposal, but that's fine since it's not a perf concern here. I'll update my code comment to point to this conversation, and recommend the utils function approach.